File: app/src/main/java/com/asfoundation/wallet/repository/TrustPasswordStore.java
package com.asfoundation.wallet.repository;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.preference.PreferenceManager;
import android.widget.Toast;
import com.asfoundation.wallet.entity.ServiceErrorException;
import com.asfoundation.wallet.entity.Wallet;
import com.asfoundation.wallet.util.KS;
import com.crashlytics.android.Crashlytics;
import com.wallet.pwd.trustapp.PasswordManager;
import io.reactivex.Completable;
import io.reactivex.Single;
import java.security.SecureRandom;
import java.util.Map;

public class TrustPasswordStore implements PasswordStore {

  private final Context context;
  private final String defaultAddress = "0x123456789";

  public TrustPasswordStore(Context context) {
    this.context = context;

    migrate();
  }

  private void migrate() {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
      return;
    }
    SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
    Map<String, ?> passwords = pref.getAll();
    for (String key : passwords.keySet()) {
      if (key.contains("-pwd")) {
        String address = key.replace("-pwd", "");
        try {
          KS.put(context, address.toLowerCase(), PasswordManager.getPassword(address, context));
        } catch (Exception ex) {
          Toast.makeText(context, "Could not process passwords.", Toast.LENGTH_LONG)
              .show();
          ex.printStackTrace();
        }
      }
    }
  }

  @Override public Single<String> getPassword(Wallet wallet) {
    return Single.fromCallable(() -> {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        return new String(KS.get(context, wallet.address));
      } else {
        return PasswordManager.getPassword(wallet.address, context);
      }
    })
        .onErrorResumeNext(throwable -> getPasswordFallBack(wallet.address));
  }

  @Override public Completable setPassword(String address, String password) {
    return Completable.fromAction(() -> {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        KS.put(context, address, password);
      } else {
        try {
          PasswordManager.setPassword(address, password, context);
        } catch (Exception e) {
          throw new ServiceErrorException(ServiceErrorException.KEY_STORE_ERROR);
        }
      }
    });
  }

  @Override public Single<String> generatePassword() {
    return Single.fromCallable(() -> {
      byte bytes[] = new byte[256];
      SecureRandom random = new SecureRandom();
      random.nextBytes(bytes);
      return new String(bytes);
    });
  }

  @Override public Completable setBackUpPassword(String masterPassword) {
    return setPassword(defaultAddress, masterPassword);
  }

  private Single<String> getPasswordFallBack(String walletAddress) {
    return Single.fromCallable(() -> {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        try {
          return new String(KS.get(context, defaultAddress));
        } catch (Exception ex) {
          Crashlytics.logException(ex);
          throw new ServiceErrorException(ServiceErrorException.KEY_STORE_ERROR,
              "Failed to get the password from the store.");
        }
      } else {
        try {
          return PasswordManager.getPassword(defaultAddress, context);
        } catch (Exception ex) {
          Crashlytics.logException(ex);
          throw new ServiceErrorException(ServiceErrorException.KEY_STORE_ERROR,
              "Failed to get the password from the password manager.");
        }
      }
    })
        .flatMap(password -> setPassword(walletAddress, password).andThen(Single.just(password)));
  }
}

File: app/src/main/java/com/asfoundation/wallet/repository/TrustPasswordStore.kt
package com.asfoundation.wallet.repository

import android.content.Context
import android.os.Build
import android.preference.PreferenceManager
import android.widget.Toast
import com.appcoins.wallet.core.utils.jvm_common.Logger
import com.asfoundation.wallet.entity.ServiceErrorException
import com.asfoundation.wallet.util.KS
import com.asfoundation.wallet.util.KS.ANDROID_KEY_STORE
import com.wallet.pwd.trustapp.PasswordManager
import dagger.hilt.android.qualifiers.ApplicationContext
import io.reactivex.Completable
import io.reactivex.Single
import it.czerwinski.android.hilt.annotations.BoundTo
import java.io.IOException
import java.security.KeyStore
import java.security.KeyStoreException
import java.security.NoSuchAlgorithmException
import java.security.SecureRandom
import java.security.cert.CertificateException
import java.util.*
import javax.inject.Inject

@BoundTo(supertype = PasswordStore::class)
class TrustPasswordStore @Inject constructor(@ApplicationContext private val context: Context,
                                             private val logger: Logger
) :
    PasswordStore {
  companion object {
    private val TAG = TrustPasswordStore::class.java.simpleName
    private const val DEFAULT_WALLET = "0x123456789"
  }

  init {
    migrate()
  }

  private fun migrate() {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
      return
    }
    val pref =
        PreferenceManager.getDefaultSharedPreferences(context)
    val passwords = pref.all
    for (key in passwords.keys) {
      if (key.contains("-pwd")) {
        val address = key.replace("-pwd", "")
        try {
          KS.put(context, address.toLowerCase(), PasswordManager.getPassword(address, context))
        } catch (ex: Exception) {
          Toast.makeText(context, "Could not process passwords.", Toast.LENGTH_LONG)
              .show()
          ex.printStackTrace()
        }
      }
    }
  }

  override fun getPassword(address: String): Single<String> {
    return Single.fromCallable {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        return@fromCallable String(KS.get(context, address))
      } else {
        return@fromCallable PasswordManager.getPassword(address, context)
      }
    }
        .onErrorResumeNext { throwable: Throwable? ->
          logError(throwable)
          getPasswordFallBack(address)
        }
  }

  override fun setPassword(address: String, password: String): Completable {
    return Completable.fromAction {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        KS.put(context, address, password)
      } else {
        try {
          PasswordManager.setPassword(address, password, context)
        } catch (e: Exception) {
          val exception = ServiceErrorException(ServiceErrorException.KEY_STORE_ERROR)
          logger.log(TAG, exception)
          throw exception
        }
      }
    }
  }

  override fun generatePassword(): Single<String> {
    return Single.fromCallable {
      val bytes = ByteArray(256)
      val random = SecureRandom()
      random.nextBytes(bytes)
      String(bytes)
    }
  }

  override fun setBackUpPassword(masterPassword: String): Completable {
    return setPassword(DEFAULT_WALLET, masterPassword)
  }

  private fun getPasswordFallBack(walletAddress: String): Single<String> {
    return Single.fromCallable {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        try {
          return@fromCallable String(KS.get(context, DEFAULT_WALLET))
        } catch (ex: Exception) {
          logger.log(TAG, ex.message, ex)
          val exception = ServiceErrorException(ServiceErrorException.KEY_STORE_ERROR,
              "Failed to get the password from the store.")
          logError(exception)
          throw exception
        }
      } else {
        try {
          return@fromCallable PasswordManager.getPassword(DEFAULT_WALLET, context)
        } catch (ex: Exception) {
          logger.log(TAG, ex.message, ex)
          val exception = ServiceErrorException(ServiceErrorException.KEY_STORE_ERROR,
              "Failed to get the password from the password manager.")
          logError(exception)
          throw exception
        }
      }
    }
        .flatMap { password: String ->
          setPassword(walletAddress, password).andThen(
              Single.just(password))
        }
  }

  private fun logError(t: Throwable?) {
    logger.log(TAG, t?.message, t)
    try {
      val keyStore = KeyStore.getInstance(ANDROID_KEY_STORE)
      keyStore.load(null)
      val strBuilder = StringBuilder()
      strBuilder.append("List of alias available in the keystore: ")
      val enumeration: Enumeration<String> = keyStore.aliases()
      while (enumeration.hasMoreElements()) {
        val alias: String = enumeration.nextElement()
        strBuilder.append("[$alias] ")
      }
      logger.log(TAG, strBuilder.toString())
    } catch (e: Exception) {
      when (e) {
        is KeyStoreException -> {
          logger.log(TAG, "Failed to get Android keystore or aliases from keystore", e)
        }
        is CertificateException,
        is IOException,
        is NoSuchAlgorithmException -> {
          logger.log(TAG, "Failed to load keystore", e)
        }
        else -> {
          logger.log(TAG, "Failed for unknown reason", e)
          throw e
        }
      }
    }
  }
}